core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -997,84 +1008,35 @@ private boolean shouldPerformTransactionOperation(StoreStatus storeStatus, Predi
     * @param flowableHandler callback handler that actually should subscribe to the underlying store
     * @param <K> key type
     * @param <V> value type
     * @return a stage that when complete will contain how many write operations were done
     */
-   private <K, V> CompletionStage<Long> batchOperation(TxInvocationContext<AbstractCacheTransaction> ctx,
-         BiPredicate<WriteCommand, Object> commandKeyPredicate, HandleFlowables<K, V> flowableHandler) {
+   private <K, V> CompletionStage<Long> batchOperation(Flowable<MVCCEntry<K, V>> mvccEntryFlowable, InvocationContext ctx,
+         HandleFlowables<K, V> flowableHandler) {
       return Single.using(
             this::acquireReadLock,
             ignore -> {
                checkStoreAvailability();
                if (trace) {
                   log.trace("Writing batch to stores");
                }
-               int size = stores.size();
-               if (size == 0) {
-                  return Single.just(0L);
-               }
-
-               Flowable<MVCCEntry<K, V>> mvccEntryFlowable = Flowable.fromIterable(ctx.getCacheTransaction().getAllModifications())
-                     .filter(writeCommand -> !writeCommand.hasAnyFlag(FlagBitSets.SKIP_CACHE_STORE))
-                     .concatMap(writeCommand -> {
-                        if (writeCommand instanceof DataWriteCommand) {
-                           Object key = ((DataWriteCommand) writeCommand).getKey();
-                           MVCCEntry<K, V> entry = acquireKeyFromContext(ctx, writeCommand, key, commandKeyPredicate);
-                           return entry != null ? Flowable.just(entry) : Flowable.empty();
-                        } else {
-                           // Assume multiple key command
-                           return Flowable.fromIterable(writeCommand.getAffectedKeys())
-                                 .concatMapMaybe(key -> {
-                                    MVCCEntry<K, V> entry = acquireKeyFromContext(ctx, writeCommand, key, commandKeyPredicate);
-                                    // We use an empty Flowable to symbolize a miss - which is filtered by ofType just below
-                                    return entry != null ? Maybe.just(entry) : Maybe.empty();
-                                 });
-                        }
-                     });
-               // We split the original source into possibly multiple, depending upon how many stores there are
-               ConnectableFlowable<MVCCEntry<K, V>> connectableFlowable = mvccEntryFlowable.publish();
-
-               // We may have multiple stores - don't publish anything until this value goes back to 0
-               AtomicInteger pendingConnects = new AtomicInteger(size);
-               if (trace) {
-                  log.tracef("Will wait for %s subscriptions of stores", size);
-               }
 
                return Flowable.fromIterable(stores)
```

Comment:
Wiith the various `if (segmented)` and `if (shared)` branches, this is now very difficult to read. Can we divide this further so that we have a `createWriteFlowable(Flowable<MarshallableEntry<K, V>> entryWriteFlowable, boolean shared, boolean segmented)` and equivalent `createRemoveFlowable` methods, containing their respective trace logs.

Reply:
Sure